doc/design/0001_customization_points.md

```diff
@@ -0,0 +1,51 @@
+# Customization points for type handling
+
+* Proposal: FFIG-0001
+* Author: Jonathan B Coe
+* Reviewer: Unassigned
+* Status: Draft
+
+# Introduction
+
+We want to be able to specify custom binding generation for user-defined types
+so that they can be handled in unusual or target-language-specific ways.
+
+# Motivation
+
+FFIG does not (yet) support generating bindings for C++ library types, nor does
+FFIG allow a user to specify how a user-defined-type should be exposed: there
+is only FFIG's way. Solving both of these problems in one go would be useful.
+
+A mechanism for specifying customisation points would allow users to specify
+how types should be handled and could be used to supply
+target-langauge-specific bindings for standard library types like
+`std::vector<double>`. Such a mechanism could be used to support the C++
+Standard Library, Abseil and Boost [FIXME(jbcoe):References needed].
```

Comment:
I think it's worth describing the different scenarios where we might encounter these types while generating bindings with ffig, for example:

*Types with language-native equivalents*
- A type with ffig bindings accepts parameters of custom / library types that have parallels in the binding language (e.g. a function accepting a `std::vector<T>` could be provided with a suitably typed Python `list` from the Python side. This would (almost certainly) involve _copying_ the data from the list into a C++ vector.
- A type with ffig bindings returns a value of custom / library type that has a parallel in the binding language (e.g. a function returning a `std::map<K, V>` which could be presented to the Python bindings as a `dict`-like object (possibly _without_ copying of the contents).

In these cases, we want to generate additional code to handle this translation. We'd need to match the type in the function signature, determine if it's a parameter or a return value (trivial) and generate the appropriate code by calling through a customisation point. We should enumerate the parameters of the custom generators to define this interface.

*Types without language-native equivalents*

In these cases, the usual ffig approach of wrapping in a pointer is likely the way to go, but it's possible that an alternative customisation is required. I would see this approach as a way of _changing_ the interface, e.g. if generating bindings for a language where objects, functions and handles / pointers don't map well onto the features of that language, or where the idiomatic way to do things in that language looks very different from C++. I think we'd need to give some examples of where we think this is useful and what kind of transformation we have in mind.